### ISSUE TITLE
pre-commit hook doesn't work with gitbutler

### ISSUE THREAD

**Issue by Cellou404** [gitbutlerapp/gitbutler#5735]
### Version

0.1

### Operating System

Linux

### Distribution Method

AppImage (Linux)

### Describe the issue

Hi, 
It's been a while that I'm using **gitbutler** for my own projects. But now, I'm working for a company and there are a bit of configuration in all our starter template. In this case, I will name **pre-commit**.
When I make a commit for changes, `pre-commit` should check my code validating the commit. But when I use **gitbutler**, no checks are made and all commit pass.
I think it would be a great thing if **gitbutler** can consider developers `pre-commit` settings.

### Expected behavior

**Gitbutler** should consider developers `pre-commit` settings so their code are checked and validated before accepting the commit.

**Comment by Byron** [comment]
Thanks for reporting!

I also think it should just respect hooks by default, but until this is the default, for now it needs to be activated on a per-project basis.

<img width="943" alt="Screenshot 2024-12-04 at 13 46 16" src="https://github.com/user-attachments/assets/007e69be-a361-4f0e-af66-706cfbe1c24c">

Does that work for you? It's known that some commit-hooks don't work as they should with GitButler as it forgoes the staging area, going straight to creating a new tree for the commit, which may prevent the hook to see what changed.

**Comment by Ebsan** [comment]
I just ran into this problem. I installed pre-commit and have a simple hook setup to look for debugging print statements. I set "Run commit hooks" to true but it still doesn't seem to run them. Doing a `git commit` from the cli works as expected (commit is rejected because of pre-commit hook.). Are there more settings we have to enable to get it to work?

I don't know if this matters, but we're using the pre-commit project at https://pre-commit.com/.

**Comment by Byron** [comment]
Enabling to run hooks per project should be all it takes.

https://github.com/gitbutlerapp/gitbutler/blob/78fb6190b7dbb8e904bef97b9aa59113a6cfe278/crates/gitbutler-branch-actions/src/virtual.rs#L721-L744

However, can it be that `pre-commit` hooks need the staging area to work? That's a typical caveat with the way GitButler works, as it bypasses the index entirely at the moment.

**Comment by Cellou404** [comment]
Hi @Byron 

Activating `Run commit hooks` is now blocking the commits but not with the appropriate error messages.

**GitButler**

![Capture d’écran du 2024-12-05 10-16-15](https://github.com/user-attachments/assets/c750ccaf-84bc-45c0-8c42-5f6cf7c90737)

**CLI**

Here's what I get when I try with cli

![Capture d’écran du 2024-12-05 10-17-52](https://github.com/user-attachments/assets/02f7e950-2a53-450a-95c0-98f1c8dfb220)

**Comment by Byron** [comment]
Thanks a lot for sharing!

Probably this is why it's behind an experimental feature toggle still, it appears that `stdout` of the hook is empty entirely. Maybe there is something on `stderr`, but that's not passed along nor logged.

I think it's easy enough to change that, I will give it a shot.

**Comment by Cellou404** [comment]
Thank you @Byron. Let me know when it's done.

**Comment by Ebsan** [comment]
> However, can it be that `pre-commit` hooks need the staging area to work? That's a typical caveat with the way GitButler works, as it bypasses the index entirely at the moment.

Yeah, the hook is looking at staging. Are there any plans to enable this? Or is there a workaround for the scripts to look somewhere else in the meantime?

**Comment by Byron** [comment]
> Yeah, the hook is looking at staging. Are there any plans to enable this? Or is there a workaround for the scripts to look somewhere else in the meantime?

I think that would require thoroughly rethinking how GitButler works. Personally, I think it should be possible to 'stage' GitButler commits so that hooks work as expected.

**Comment by Cellou404** [comment]
@Byron

**Comment by Caleb-T-Owens** [comment]
Hi @Cellou404, I believe @mtsgrd is currently working on this issue

**Comment by mtsgrd** [comment]
This issue has now been fixed as of the latest release. It would be great if someone on this issue could give it a try and let me know if things are working as expected?

**Comment by aquarhead** [comment]
Not sure in what environment the hooks are being executed, but it's not using the right PATH

```
Check format.............................................................Failed - hook id: rubyfmt - exit code: 1 Executable `rubyfmt` not found Check lint
```

Example `.pre-commit-config.yaml`:

```
repos:
- repo: local
  hooks:
  - id: rubyfmt
    name: Check format
    entry: rubyfmt -c app/
    language: system
    always_run: true
    pass_filenames: false
```

**Comment by Byron** [comment]
@aquarhead Does this work if GitButler is launched directly from a terminal? On MacOS this would be `/Applications/GitButler.app/Contents/MacOS/gitbutler-tauri`. Thank you.

**Comment by aquarhead** [comment]
> [@aquarhead](https://github.com/aquarhead) Does this work if GitButler is launched directly from a terminal? On MacOS this would be `/Applications/GitButler.app/Contents/MacOS/gitbutler-tauri`. Thank you.

It looks like it runs correctly when launched from my usual terminal

![Image](https://github.com/user-attachments/assets/67ba391e-77da-415c-90f9-3e081f5a060c)

**Comment by Byron** [comment]
Great, thank you!

This means we'd really want a way to help users with running hooks and external programs in the same environment as they are used to when launching from their login shells. This is a complex topic that isn't easily solved, but maybe it can be supported if we would make it easy to alter the application environment, for instance with a copy of the output of `env` which can then be adjusted further in a simple text field.

**Comment by Byron** [comment]
This issue is about hooks, and I think authentication is unrelated to that. If there is an issue in that regard, would you mind opening a new one? Thank you 🙏.

**Comment by aquarhead** [comment]
> This issue is about hooks, and I think authentication is unrelated to that. If there is an issue in that regard, would you mind opening a new one? Thank you 🙏.

I don't have any issue with auth, just wondering if it already uses "the right env" or maybe it can be "fixed" in the same way.

Anyways, I really enjoy using this tool for some projects and looking forward to a fix for hooks.

**Comment by Cellou404** [comment]
It's now working fine 👌

A million thanks to you @Byron @Caleb-T-Owens @mtsgrd  and all the team.
I'm enjoying using this tool 😇

![Image](https://github.com/user-attachments/assets/2ae3d686-a646-4b22-b1c0-b83231446cd7)

**Comment by mtsgrd** [comment]
I'm glad to hear! Now we just need a better way of displaying the results.

**Comment by Cellou404** [comment]
> I'm glad to hear! Now we just need a better way of displaying the results.

Yeah, it will be a greate thing

**Comment by stevelw** [comment]
I have setup husky hooks and the pre-commit (only one added) isn't running.

Are there any docs on how this feature is expected to be setup?

**Comment by Byron** [comment]
It should just work actually. But it looks like something is off now, as `core.hooksPath` also doesn't seem to work (anymore).

**Comment by b2m9** [comment]
Can confirm, pre-commit hooks aren't working. I assume #8090 is the ticket for that, @Byron?

**Comment by Byron** [comment]
Thanks for the reminder - I have been waiting for a new release for `git2-hooks` but decided to just go ahead now with a Git dependency.

Please see #8709, a PR that should land soon, and will probably make it into today's patch release as well.
